Forge.TreeWalker/src/ForgeSchemaValidator.cs

```diff
@@ -0,0 +1,284 @@
+        /// </summary>
+        /// <param name="childRules">The rules to be included in the parent rules</param>
+        /// <param name="parentRules">The parent rules to absorb the child rules</param>
+        /// <param name="referenceUri">The address of childRules</param>
+        /// <returns>The result of schema combination. ErrorMessage would be set if it throw exceptions</returns>
+        public static JSchema GetLinkedJSchemaRules(string childRules, string parentRules, string referenceUri, out string errorMessage)
+        {
+            try
+            {
+                JSchemaPreloadedResolver resolver = new JSchemaPreloadedResolver();
+                resolver.Add(new Uri(referenceUri), childRules);
+                errorMessage = "";
+                return JSchema.Parse(parentRules, resolver);
+            }
+            catch(Exception e)
+            {
+                errorMessage = e.Message;
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// The validate task that validate the input schema with custom rules in string.
+        /// </summary>
+        /// <param name="schema">The schema to be validated</param>
+        /// <param name="rules">The rules used to validate input schemas</param>
+        /// <returns>The result of schema validation. The errorList would contain error message if validation fails</returns>
+        public static bool ValidateSchema(ForgeTree schema, string rules, out List<string> errorList)
+        {
+            return Validate(new List<JObject> { SerializeForgeTree(schema) }, JSchema.Parse(rules), out errorList);
+        }
+        /// <summary>
+        /// The validate task that validate the input schema with custom rules in string.
+        /// </summary>
+        /// <param name="schema">The schema to be validated</param>
+        /// <param name="rules">The rules used to validate input schemas</param>
+        /// <returns>The result of schema validation. The errorList would contain error message if validation fails</returns>
+        public static bool ValidateSchema(ForgeTree schema, JSchema rules, out List<string> errorList)
+        {
+            return Validate(new List<JObject> { SerializeForgeTree(schema) }, rules, out errorList);
+        }
+        /// <summary>
+        /// The validate task that validate multiple input schemas with custom rules in string.
+        /// </summary>
+        /// <param name="schemas">The schemas to be validated</param>
+        /// <param name="rules">The rules used to validate input schemas</param>
+        /// <returns>The result of schema validation. The errorList would contain error message if validation fails</returns>
+        public static bool ValidateSchemas(Dictionary<string, ForgeTree> schemas, string rules, out List<string> errorList)
+        {
+            return Validate(ConvertDictionaryToForgeTreeList(schemas), JSchema.Parse(rules), out errorList);
```

Comment:
>Validate(ConvertDictionaryToForgeTreeList(schemas) [](http://example.com/codeflow?start=19&length=50)

When users use Dictionary and/or have separate files, I've seen two ways they organize their files.

1. Inside each schema file, they have the TreeName mapped to the ForgeTree, i.e. a dictionary.

In this case, the team has a custom rules file that handles the dictionary. So the string schema file they validate the rules for is a dictionary (and must be a dictionary). If we break out the dictionary into a list of ForgeTrees, the custom rules will not work.

Let's either assume this behavior (custom rules is for dictionary if passing a dictionary), or allow the user to specify if they want to validate their schemas one-by-one or all together.

2. Inside each schema file, they do not have a TreeName. They only have a ForgeTree.

In this case, they team is either using the default rules file or their custom rules file does not include an added dictionary. So it would make sense to validate each of their ForgeTrees in a list in this case.

We can discuss offline which route to go here.

Reply:
Just looked into how we are using this in one of our applications. We have multiple schema files, each containing a dictionary. We first deserialize each file into a dictionary, and add it to a combined dictionary. We then serialize the combined dictionary into a single string schema.

I'm leaning towards covering both ways with a flag for validateAsDictionary.

This would be used for string, Dictionary, path, or directory cases.

For directory in particular, we should also do what I called out above when validateAsDictionary is true. Because if the files are each a dictionary, we need to combine them all before running the validation.